[Bug]: [upgrade] rootcoord panic: invalid channel name when upgrade from v2.2.5 to v2.2.7

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Environment

```markdown
- Milvus version: v2.2.7
- Deployment mode(standalone or cluster): cluster
- MQ type(rocksmq, pulsar or kafka):    
- SDK version(e.g. pymilvus v2.0.0rc2):
- OS(Ubuntu or CentOS): 
- CPU/Memory: 
- GPU: 
- Others:
```


### Current Behavior

1. deploy milvus with v2.2.0 image and run go-sdk v2.2.x  branch test
2. upgrade milvus to v2.2.1 and run go-sdk v2.2.x  branch test
3. upgrade milvus to v2.2.2 and run go-sdk v2.2.x  branch test
4. upgrade milvus to v2.2.3 and run go-sdk v2.2.x  branch test
5. upgrade milvus to v2.2.4 and run go-sdk v2.2.x  branch test
6. upgrade milvus to v2.2.5 and run go-sdk v2.2.x  branch test
7. upgrade milvus to v2.2.7 and rootcoord panic, milvus is not healthy

panic log:
```
[2023/05/01 09:35:48.664 +00:00] [INFO] [rootcoord/dml_channels.go:176] ["init dml channels"] [prefix=go-sdk-rootcoord-dml] [num=84]
[2023/05/01 09:35:48.664 +00:00] [INFO] [rootcoord/timeticksync.go:121] ["recover physical channels"] [collID=441073883353124913] ["physical channels"="[go-sdk-rootcoord-dml_34,go-sdk-rootcoord-dml_35]"]
[2023/05/01 09:35:48.664 +00:00] [INFO] [rootcoord/timeticksync.go:121] ["recover physical channels"] [collID=441075963583729842] ["physical channels"="[go-sdk-rootcoord-dml_46,go-sdk-rootcoord-dml_47]"]
[2023/05/01 09:35:48.664 +00:00] [ERROR] [rootcoord/dml_channels.go:281] ["invalid channel name"] [chanName=go-sdk-rootcoord-dml_84] [stack="github.com/milvus-io/milvus/internal/rootcoord.(*dmlChannels).addChannels\n\t/go/src/github.com/milvus-io/milvus/internal/rootcoord/dml_channels.go:281\ngithub.com/milvus-io/milvus/internal/rootcoord.newTimeTickSync\n\t/go/src/github.com/milvus-io/milvus/internal/rootcoord/timeticksync.go:120\ngithub.com/milvus-io/milvus/internal/rootcoord.(*Core).initInternal\n\t/go/src/github.com/milvus-io/milvus/internal/rootcoord/root_coord.go:445\ngithub.com/milvus-io/milvus/internal/rootcoord.(*Core).Init.func2\n\t/go/src/github.com/milvus-io/milvus/internal/rootcoord/root_coord.go:518\nsync.(*Once).doSlow\n\t/usr/local/go/src/sync/once.go:68\nsync.(*Once).Do\n\t/usr/local/go/src/sync/once.go:59\ngithub.com/milvus-io/milvus/internal/rootcoord.(*Core).Init\n\t/go/src/github.com/milvus-io/milvus/internal/rootcoord/root_coord.go:517\ngithub.com/milvus-io/milvus/internal/distributed/rootcoord.(*Server).init\n\t/go/src/github.com/milvus-io/milvus/internal/distributed/rootcoord/service.go:212\ngithub.com/milvus-io/milvus/internal/distributed/rootcoord.(*Server).Run\n\t/go/src/github.com/milvus-io/milvus/internal/distributed/rootcoord/service.go:142\ngithub.com/milvus-io/milvus/cmd/components.(*RootCoord).Run\n\t/go/src/github.com/milvus-io/milvus/cmd/components/root_coord.go:57\ngithub.com/milvus-io/milvus/cmd/roles.runComponent[...].func1\n\t/go/src/github.com/milvus-io/milvus/cmd/roles/roles.go:102"]
panic: invalid channel name: go-sdk-rootcoord-dml_84

goroutine 239 [running]:
github.com/milvus-io/milvus/internal/rootcoord.(*dmlChannels).addChannels(0xc001fb4a00, {0xc001fa1d60?, 0x2, 0xc0014fccf0?})
    /go/src/github.com/milvus-io/milvus/internal/rootcoord/dml_channels.go:282 +0x267
github.com/milvus-io/milvus/internal/rootcoord.newTimeTickSync({0x4209ef0?, 0xc00088c7c0}, 0x41, {0x7fdafcfa73a0, 0xc0014fccf0}, 0xc0022f3f90?)
    /go/src/github.com/milvus-io/milvus/internal/rootcoord/timeticksync.go:120 +0x15f
github.com/milvus-io/milvus/internal/rootcoord.(*Core).initInternal(0xc000ae2a80)
    /go/src/github.com/milvus-io/milvus/internal/rootcoord/root_coord.go:445 +0x1ef
github.com/milvus-io/milvus/internal/rootcoord.(*Core).Init.func2()
    /go/src/github.com/milvus-io/milvus/internal/rootcoord/root_coord.go:518 +0x26
sync.(*Once).doSlow(0xc000ae2a80?, 0x57fba60?)
    /usr/local/go/src/sync/once.go:68 +0xc2
sync.(*Once).Do(...)
    /usr/local/go/src/sync/once.go:59
github.com/milvus-io/milvus/internal/rootcoord.(*Core).Init(0xc000ae2a80)
    /go/src/github.com/milvus-io/milvus/internal/rootcoord/root_coord.go:517 +0xac
github.com/milvus-io/milvus/internal/distributed/rootcoord.(*Server).init(0xc00089e370)
    /go/src/github.com/milvus-io/milvus/internal/distributed/rootcoord/service.go:212 +0x5ee
github.com/milvus-io/milvus/internal/distributed/rootcoord.(*Server).Run(0xc0007aff00?)
    /go/src/github.com/milvus-io/milvus/internal/distributed/rootcoord/service.go:142 +0x25
github.com/milvus-io/milvus/cmd/components.(*RootCoord).Run(0x57fba60?)
    /go/src/github.com/milvus-io/milvus/cmd/components/root_coord.go:57 +0x2f
github.com/milvus-io/milvus/cmd/roles.runComponent[...].func1()
    /go/src/github.com/milvus-io/milvus/cmd/roles/roles.go:102 +0x18e
created by github.com/milvus-io/milvus/cmd/roles.runComponent[...]
    /go/src/github.com/milvus-io/milvus/cmd/roles/roles.go:85 +0x18a
```

### Expected Behavior

_No response_

### Steps To Reproduce

```markdown
1. deploy milvus with some version
2. run go-sdk test by devops cluster argo `go-sdk-test-release` workflow template
3. upgrade milvus version
```


### Milvus Log

rootcoord panic log: 
[rootcoord.log](https://github.com/milvus-io/milvus/files/11364805/rootcoord.log)

pods in devops cluster and chaos-testing ns:
```
go-sdk-milvus-datacoord-759fc7c97f-9lhr8                          0/1     Running            6 (3m3s ago)       21m
go-sdk-milvus-datanode-5cfcb689b5-m4vg2                           1/1     Running            0                  4d3h
go-sdk-milvus-datanode-7ff84dc75-ct9wj                            0/1     Running            7 (18s ago)        21m
go-sdk-milvus-indexcoord-5c9cbf5679-tfl2v                         0/1     Running            6 (3m3s ago)       21m
go-sdk-milvus-indexnode-7667999-6xnv8                             1/1     Running            0                  21m
go-sdk-milvus-proxy-5795546758-tw4vw                              0/1     Running            7 (7s ago)         21m
go-sdk-milvus-proxy-bd579777b-4x4c5                               1/1     Running            0                  4d3h
go-sdk-milvus-querycoord-64c499d8ff-dftf6                         0/1     Running            7 (4s ago)         21m
go-sdk-milvus-querynode-58f46cf444-vsj85                          1/1     Running            0                  21m
go-sdk-milvus-rootcoord-7995b9cfc5-wt756                          0/1     CrashLoopBackOff   8 (4m22s ago)      21m
```

### Anything else?

_No response_

--- a/internal/querycoordv2/server.go
+++ b/internal/querycoordv2/server.go
@@ -626,7 +626,10 @@ func (s *Server) watchNodes(revision int64) {
 				)
 				s.nodeMgr.Add(session.NewNodeInfo(nodeID, addr))
 				s.nodeUpEventChan <- nodeID
-				s.notifyNodeUp <- struct{}{}
+				select {
+				case s.notifyNodeUp <- struct{}{}:
+				default:
+				}
 
 			case sessionutil.SessionUpdateEvent:
 				nodeID := event.Session.ServerID

--- a/internal/rootcoord/dml_channels.go
+++ b/internal/rootcoord/dml_channels.go
@@ -20,14 +20,16 @@ import (
 	"container/heap"
 	"context"
 	"fmt"
+	"strconv"
+	"strings"
 	"sync"
 
 	"github.com/milvus-io/milvus/internal/metrics"
 
-	"go.uber.org/zap"
-
 	"github.com/milvus-io/milvus/internal/log"
 	"github.com/milvus-io/milvus/internal/mq/msgstream"
+	"github.com/milvus-io/milvus/internal/util/typeutil"
+	"go.uber.org/zap"
 )
 
 type dmlMsgStream struct {
@@ -308,4 +310,33 @@ func (d *dmlChannels) removeChannels(names ...string) {
 
 func genChannelName(prefix string, idx int64) string {
 	return fmt.Sprintf("%s_%d", prefix, idx)
+}
+
+func parseChannelNameIndex(channeName string) int {
+	index := strings.LastIndex(channeName, "_")
+	if index < 0 {
+		log.Error("invalid channel name", zap.String("chanName", channeName))
+		panic("invalid channel name: " + channeName)
+	}
+	index, err := strconv.Atoi(channeName[index+1:])
+	if err != nil {
+		log.Error("invalid channel name", zap.String("chanName", channeName), zap.Error(err))
+		panic("invalid channel name: " + channeName)
+	}
+	return index
+}
+
+func getNeedChanNum(setNum int, chanMap map[typeutil.UniqueID][]string) int {
+	// find the largest number of current channel usage
+	maxChanUsed := setNum
+	for _, chanNames := range chanMap {
+		for _, chanName := range chanNames {
+			index := parseChannelNameIndex(chanName)
+			if maxChanUsed < index+1 {
+				maxChanUsed = index + 1
+			}
+		}
+	}
+
+	return maxChanUsed
 }

--- a/internal/rootcoord/timeticksync.go
+++ b/internal/rootcoord/timeticksync.go
@@ -110,10 +110,10 @@ func (c *chanTsMsg) getTimetick(channelName string) typeutil.Timestamp {
 func newTimeTickSync(ctx context.Context, sourceID int64, factory msgstream.Factory, chanMap map[typeutil.UniqueID][]string) *timetickSync {
 	// if the old channels number used by the user is greater than the set default value currently
 	// keep the old channels
-	defaultChanNum := getNeedChanNum(int(Params.RootCoordCfg.DmlChannelNum), chanMap)
+	channelNum := getNeedChanNum(int(Params.RootCoordCfg.DmlChannelNum), chanMap)
 
 	// initialize dml channels used for insert
-	dmlChannels := newDmlChannels(ctx, factory, Params.CommonCfg.RootCoordDml, int64(defaultChanNum))
+	dmlChannels := newDmlChannels(ctx, factory, Params.CommonCfg.RootCoordDml, int64(channelNum))
 
 	// recover physical channels for all collections
 	for collID, chanNames := range chanMap {

--- a/internal/rootcoord/timeticksync_test.go
+++ b/internal/rootcoord/timeticksync_test.go
@@ -118,3 +118,80 @@ func Test_ttHistogram_get(t *testing.T) {
 	assert.Equal(t, typeutil.ZeroTimestamp, h.get("ch1"))
 	assert.Equal(t, typeutil.ZeroTimestamp, h.get("ch2"))
 }
+
+func TestTimetickSyncWithExistChannels(t *testing.T) {
+	ctx := context.Background()
+	sourceID := int64(100)
+
+	factory := dependency.NewDefaultFactory(true)
+
+	//chanMap := map[typeutil.UniqueID][]string{
+	//	int64(1): {"rootcoord-dml_0"},
+	//}
+
+	Params.RootCoordCfg.DmlChannelNum = 2
+	Params.CommonCfg.RootCoordDml = "rootcoord-dml"
+	Params.CommonCfg.RootCoordDelta = "rootcoord-delta"
+	chans := map[UniqueID][]string{}
+
+	chans[UniqueID(100)] = []string{"rootcoord-dml_4", "rootcoord-dml_8"}
+	chans[UniqueID(102)] = []string{"rootcoord-dml_2", "rootcoord-dml_9"}
+	ttSync := newTimeTickSync(ctx, sourceID, factory, chans)
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+	t.Run("sendToChannel", func(t *testing.T) {
+		defer wg.Done()
+		ttSync.sendToChannel()
+
+		ttSync.sess2ChanTsMap[1] = nil
+		ttSync.sendToChannel()
+
+		msg := &internalpb.ChannelTimeTickMsg{
+			Base: &commonpb.MsgBase{
+				MsgType: commonpb.MsgType_TimeTick,
+			},
+		}
+		ttSync.sess2ChanTsMap[1] = newChanTsMsg(msg, 1)
+		ttSync.sendToChannel()
+	})
+
+	wg.Add(1)
+	t.Run("assign channels", func(t *testing.T) {
+		defer wg.Done()
+		channels := ttSync.getDmlChannelNames(int(4))
+		assert.Equal(t, channels, []string{"rootcoord-dml_0", "rootcoord-dml_1", "rootcoord-dml_3", "rootcoord-dml_5"})
+
+		channels = ttSync.getDmlChannelNames(int(4))
+		assert.Equal(t, channels, []string{"rootcoord-dml_6", "rootcoord-dml_7", "rootcoord-dml_0", "rootcoord-dml_1"})
+	})
+
+	// test get new channels
+
+}
+
+func TestTimetickSyncInvalidName(t *testing.T) {
+	ctx := context.Background()
+	sourceID := int64(100)
+
+	factory := dependency.NewDefaultFactory(true)
+
+	//chanMap := map[typeutil.UniqueID][]string{
+	//	int64(1): {"rootcoord-dml_0"},
+	//}
+
+	Params.RootCoordCfg.DmlChannelNum = 2
+	Params.CommonCfg.RootCoordDml = "rootcoord-dml"
+	Params.CommonCfg.RootCoordDelta = "rootcoord-delta"
+	chans := map[UniqueID][]string{}
+	chans[UniqueID(100)] = []string{"rootcoord-dml4"}
+	assert.Panics(t, func() {
+		newTimeTickSync(ctx, sourceID, factory, chans)
+	})
+
+	chans = map[UniqueID][]string{}
+	chans[UniqueID(102)] = []string{"rootcoord-dml_a"}
+	assert.Panics(t, func() {
+		newTimeTickSync(ctx, sourceID, factory, chans)
+	})
+}

--- a/internal/rootcoord/util.go
+++ b/internal/rootcoord/util.go
@@ -137,16 +137,3 @@ func getTravelTs(req TimeTravelRequest) Timestamp {
 func isMaxTs(ts Timestamp) bool {
 	return ts == typeutil.MaxTimestamp
 }
-
-func getNeedChanNum(setNum int, chanMap map[typeutil.UniqueID][]string) int {
-	chanNames := typeutil.NewSet[string]()
-	for _, chanName := range chanMap {
-		chanNames.Insert(chanName...)
-	}
-
-	ret := chanNames.Len()
-	if setNum > chanNames.Len() {
-		ret = setNum
-	}
-	return ret
-}

--- a/internal/rootcoord/util_test.go
+++ b/internal/rootcoord/util_test.go
@@ -150,22 +150,3 @@ func Test_isMaxTs(t *testing.T) {
 		})
 	}
 }
-
-func Test_GetNeedChanNum(t *testing.T) {
-	chanMap := map[typeutil.UniqueID][]string{
-		int64(1): {"rootcoord-dml_101"},
-		int64(2): {"rootcoord-dml_102"},
-		int64(3): {"rootcoord-dml_103"},
-	}
-
-	num := getNeedChanNum(2, chanMap)
-	assert.Equal(t, num, 3)
-
-	chanMap = map[typeutil.UniqueID][]string{
-		int64(1): {"rootcoord-dml_101", "rootcoord-dml_102"},
-		int64(2): {"rootcoord-dml_102", "rootcoord-dml_101"},
-		int64(3): {"rootcoord-dml_103", "rootcoord-dml_102"},
-	}
-	num = getNeedChanNum(2, chanMap)
-	assert.Equal(t, num, 3)
-}

--- a/tests/python_client/testcases/test_compaction.py
+++ b/tests/python_client/testcases/test_compaction.py
@@ -370,7 +370,7 @@ class TestCompactionOperation(TestcaseBase):
                 3.delete and flush (new insert)
                 4.compact
                 5.load and search
-        expected: Triggre two types compaction
+        expected: Trigger two types compaction
         """
         collection_w = self.init_collection_wrap(cf.gen_unique_str(prefix), shards_num=1)
         ids = []
@@ -545,10 +545,10 @@ class TestCompactionOperation(TestcaseBase):
         c_plans = collection_w.get_compaction_plans(check_task=CheckTasks.check_merge_compact)[0]
 
         # waiting for handoff completed and search
-        cost = 60
+        cost = 180
         start = time()
         while True:
-            sleep(5)
+            sleep(1)
             segment_info = self.utility_wrap.get_query_segment_info(collection_w.name)[0]
             if len(segment_info) != 0 and segment_info[0].segmentID == c_plans.plans[0].target:
                 log.debug(segment_info)

